File: typescript/src/server/address/validate/restApiAddressValidateRequest.ts
import { AddressConfiguration } from "../../../address/addressConfiguration";
import { RestApiFormatAttribute } from "../format/restApiFormatAttribute";
import { RestApiAdditionalOption } from "../restApiAdditionalOption";
import { RestApiAddress } from "../restApiAddress";

export type RestApiAddressValidateRequest = {
    country_iso?: string;
    datasets?: string[];
    max_suggestions?: number;
    components?: RestApiAddress;
    options?: RestApiAdditionalOption[];
    attributes?: RestApiFormatAttribute;
    layouts?: string[];
    layout_format?: string;
};

export function getAddressValidateRequestFromConfig(config: AddressConfiguration): RestApiAddressValidateRequest {
    const request: RestApiAddressValidateRequest = {options: []};
    // Country
    if (config.options.datasets && config.options.datasets.length > 0) {
        request.country_iso = config.options.datasets[0].country.iso3Code;
    }
    // Datasets
    if (config.options.datasets && config.options.datasets.length > 0) {
        request.datasets = config.options.datasets.map(dataset => dataset.datasetCode);
    }
    // Max suggestions
    if (config.options.maxSuggestions !== AddressConfiguration.defaultMaxSuggestions) {
        request.max_suggestions = config.options.maxSuggestions;
    }
    //Layout name
    if (config.options.formatLayoutName && config.options.formatLayoutName.length > 0) {
        request.layouts = [config.options.formatLayoutName];
    }
    //Layout format
    if (config.options.layoutFormat && config.options.layoutFormat.length > 0) {
        request.layout_format = config.options.layoutFormat;
    }
    if (config.options.formatLayoutName) {
        request.options!.push({ name: "flatten", value: "true" });
    }
    if (config.options.searchIntensity) {
        request.options!.push({ name: "intensity", value: config.options.searchIntensity });
    }
    if (config.options.promptSet) {
        request.options!.push({ name: "prompt_set", value: config.options.promptSet });
    }

    request.attributes = getFormatAttribute(config);

    return request;
    
}

export function getFormatAttribute(configuration: AddressConfiguration): RestApiFormatAttribute {
    const attributes: RestApiFormatAttribute = {};

    if (configuration.options.globalGeocodes && configuration.options.globalGeocodes.length > 0) {
        attributes.geocodes = configuration.options.globalGeocodes;;
    }

    if (configuration.options.premiumLocationInsights && configuration.options.premiumLocationInsights.length > 0) {
        attributes.premium_location_insight = configuration.options.premiumLocationInsights;
    }

    if (configuration.options.what3Words && configuration.options.what3Words.length > 0) {
        attributes.what3words = configuration.options.what3Words;
    }

    if (configuration.options.gbrLocationComplete && configuration.options.gbrLocationComplete.length > 0) {
        attributes.uk_location_complete = configuration.options.gbrLocationComplete;
    }

    if (configuration.options.gbrLocationEssential && configuration.options.gbrLocationEssential.length > 0) {
        attributes.uk_location_essential = configuration.options.gbrLocationEssential;
    }

    if (configuration.options.gbrGovernment && configuration.options.gbrGovernment.length > 0) {
        attributes.gbr_government = configuration.options.gbrGovernment;
    }

    if (configuration.options.gbrHealth && configuration.options.gbrHealth.length > 0) {
        attributes.gbr_health = configuration.options.gbrHealth;
    }

    if (configuration.options.gbrBusiness && configuration.options.gbrBusiness.length > 0) {
        attributes.gbr_business = configuration.options.gbrBusiness;
    }

    if (configuration.options.usaRegionalGeocodes && configuration.options.usaRegionalGeocodes.length > 0) {
        attributes.usa_regional_geocodes = configuration.options.usaRegionalGeocodes;
    }

    if (configuration.options.ausRegionalGeocodes && configuration.options.ausRegionalGeocodes.length > 0) {
        attributes.aus_regional_geocodes = configuration.options.ausRegionalGeocodes;
    }

    if (configuration.options.nzlRegionalGeocodes && configuration.options.nzlRegionalGeocodes.length > 0) {
        attributes.nzl_regional_geocodes = configuration.options.nzlRegionalGeocodes;
    }

    return attributes;

}    

File: csharp/DVSClient/DVSClient/server/address/validate/RestApiAddressValidateRequest.cs
using DVSClient.Address.Layout.Attributes;
using DVSClient.Common;
using DVSClient.Server.Address.Format;
using DVSClient.Server.Address.Search;
using Newtonsoft.Json;

namespace DVSClient.Server.Address.Validate
{
    public class RestApiAddressValidateRequest
    {
        [JsonProperty("country_iso")]
        public string? CountryIso { get; set; }

        [JsonProperty("datasets")]
        public IEnumerable<string>? Datasets { get; set; }

        [JsonProperty("max_suggestions")]
        public int? MaxSuggestions { get; set; }

        [JsonProperty("components")]
        public Address? Address { get; set; }

        [JsonProperty("options")]
        public IList<RestApiAdditionalOption>? Options { get; set; }

        [JsonProperty("attributes")]
        public RestApiFormatAttribute? Attributes { get; set; }

        [JsonProperty("layouts")]
        public IEnumerable<string>? Layouts { get; set; }

        [JsonProperty("layout_format")]
        public string? LayoutFormat { get; set; }

        public static RestApiAddressValidateRequest Using(DVSClient.Address.AddressConfiguration configuration)
        {
            var validateRequest = new RestApiAddressValidateRequest();

            // Country
            if (configuration.GetCountry() != null)
            {
                validateRequest.CountryIso = configuration.GetCountry()?.Iso3Code;
            }

            // Datasets
            if (configuration.Datasets != null && configuration.Datasets.Any())
            {
                validateRequest.Datasets = configuration.Datasets.ToList().ConvertAll(dataset => dataset.DatasetCode);
            }

            // Max suggestions
            if (configuration.MaxSuggestions != DVSClient.Address.AddressConfiguration.DefaultMaxSuggestions)
            {
                validateRequest.MaxSuggestions = configuration.MaxSuggestions;
            }

            // Layout name
            if (!string.IsNullOrEmpty(configuration.FormatLayoutName))
            {
                validateRequest.Layouts = new List<string> { configuration.FormatLayoutName };
            }

            // Layout format
            if (configuration.LayoutFormat.HasValue)
            {
                validateRequest.LayoutFormat = configuration.LayoutFormat.Value.ToString();
            }

            if (configuration.FormatLayoutName != null)
            {
                validateRequest.AddOption("flatten", true.ToString());
            }

            if (configuration.SearchIntensity.HasValue)
            {
                validateRequest.AddOption("intensity", configuration.SearchIntensity.Value.ToString());
            }

            if (configuration.PromptSet.HasValue)
            {
                validateRequest.AddOption("prompt_set", configuration.PromptSet.Value.ToString());
            }

            var formatAttributes = GetFormatAttribute(configuration);
            if (formatAttributes != null)
            {
                validateRequest.Attributes = formatAttributes;
            }

            return validateRequest;
        }

        public static RestApiFormatAttribute GetFormatAttribute(DVSClient.Address.AddressConfiguration configuration)
        {
            var attributes = new RestApiFormatAttribute();

            if (configuration.GlobalGeocodes != null && configuration.GlobalGeocodes.Any())
            {
                attributes.Geocodes = configuration.GlobalGeocodes.Select(attr => attr.GetJsonNameFromEnum<GlobalGeocodeAttribute>());
            }

            if (configuration.PremiumLocationInsights != null && configuration.PremiumLocationInsights.Any())
            {
                attributes.PremiumLocationInsight = (IEnumerable<string?>?)configuration.PremiumLocationInsights.Select(attr => attr.GetJsonNameFromEnum<PremiumLocationInsightAttribute>());
            }

            if (configuration.What3Words != null && configuration.What3Words.Any())
            {
                attributes.What3words = (IEnumerable<string?>?)configuration.What3Words.Select(attr => attr.GetJsonNameFromEnum<What3WordsAttribute>());
            }

            if (configuration.GbrLocationComplete != null && configuration.GbrLocationComplete.Any())
            {
                attributes.GbrLocationComplete = (IEnumerable<string?>?)configuration.GbrLocationComplete.Select(attr => attr.GetJsonNameFromEnum<GbrLocationCompleteAttribute>());
            }

            if (configuration.GbrLocationEssential != null && configuration.GbrLocationEssential.Any())
            {
                attributes.GbrLocationEssential = (IEnumerable<string?>?)configuration.GbrLocationEssential.Select(attr => attr.GetJsonNameFromEnum<GbrLocationEssentialAttribute>());
            }

            if (configuration.GbrGovernment != null && configuration.GbrGovernment.Any())
            {
                attributes.GbrGovernment = (IEnumerable<string?>?)configuration.GbrGovernment.Select(attr => attr.GetJsonNameFromEnum<GbrGovernmentAttribute>());
            }

            if (configuration.GbrHealth != null && configuration.GbrHealth.Any())
            {
                attributes.GbrHealth = (IEnumerable<string?>?)configuration.GbrHealth.Select(attr => attr.GetJsonNameFromEnum<GbrHealthAttribute>());
            }

            if (configuration.GbrBusiness != null && configuration.GbrBusiness.Any())
            {
                attributes.GbrBusiness = (IEnumerable<string?>?)configuration.GbrBusiness.Select(attr => attr.GetJsonNameFromEnum<GbrBusinessAttribute>());
            }

            if (configuration.UsaRegionalGeocodes != null && configuration.UsaRegionalGeocodes.Any())
            {
                attributes.UsaRegionalGeocodes = (IEnumerable<string?>?)configuration.UsaRegionalGeocodes.Select(attr => attr.GetJsonNameFromEnum<UsaRegionalGeocodeAttribute>());
            }

            if (configuration.AusRegionalGeocodes != null && configuration.AusRegionalGeocodes.Any())
            {
                attributes.AusRegionalGeocodes = (IEnumerable<string?>?)configuration.AusRegionalGeocodes.Select(attr => attr.GetJsonNameFromEnum<AusRegionalGeocodeAttribute>());
            }

            if (configuration.NzlRegionalGeocodes != null && configuration.NzlRegionalGeocodes.Any())
            {
                attributes.NzlRegionalGeocodes = (IEnumerable<string?>?)configuration.NzlRegionalGeocodes.Select(attr => attr.GetJsonNameFromEnum<NzlRegionalGeocodeAttribute>());
            }

            return attributes;
        }

        public void AddOption(string name, string value)
        {
            if (Options == null)
            {
                Options = new List<RestApiAdditionalOption>();
            }

            Options.Add(new RestApiAdditionalOption(name, value));
        }
    }
}